armi/reactor/converters/axialExpansionChanger/assemblyAxialLinkage.py

```diff
@@ -57,118 +73,187 @@ def _determineLinked(componentA, componentB):
 
+    Parameters
+    ----------
+    assem : armi.reactor.assemblies.Assembly
+        Assembly to be linked
+
     Attributes
     ----------
     a : :py:class:`Assembly <armi.reactor.assemblies.Assembly>`
         reference to original assembly; is directly modified/changed during expansion.
     linkedBlocks : dict
-        - keys = :py:class:`Block <armi.reactor.blocks.Block>`
-        - values = list of axially linked blocks; index 0 = lower linked block; index 1: upper
-          linked block.
+        Keys are blocks in the assembly. Their values are :class:`AxialLink` with
+        ``upper`` and ``lower`` attributes for the blocks potentially above and
+        below this block.
     linkedComponents : dict
-        - keys = :py:class:`Component <armi.reactor.components.component.Component>`
-        - values = list of axially linked components; index 0 = lower linked component;
-          index 1: upper linked component.
+        Keys are solid components in the assembly. Their values are :class:`AxialLink` with
+        ``upper`` and ``lower`` attributes for the solid components potentially above and
+        below this block.
     """
 
-    def __init__(self, StdAssem):
-        self.a = StdAssem
-        self.linkedBlocks = {}
+    linkedBlocks: dict[Block, AxialLink[Block]]
+    linkedComponents: dict[Component, AxialLink[Component]]
+
+    def __init__(self, assem: "Assembly"):
+        self.a = assem
+        self.linkedBlocks = self.getLinkedBlocks(assem)
         self.linkedComponents = {}
         self._determineAxialLinkage()
 
+    @classmethod
+    def getLinkedBlocks(
+        cls,
+        blocks: typing.Sequence[Block],
+    ) -> dict[Block, AxialLink[Block]]:
+        """Produce a mapping showing how blocks are linked.
+
+        Parameters
+        ----------
+        blocks : sequence of armi.reactor.blocks.Block
+            Ordered sequence of blocks from bottom to top. Could just as easily be an
+            :class:`armi.reactor.assemblies.Assembly`.
+
+        Returns
+        -------
+        dict of Block -> AxialLink
+            Dictionary where keys are individual blocks and their corresponding values point
+            to blocks above and below.
+        """
+        nBlocks = len(blocks)
+        if nBlocks:
+            return cls._getLinkedBlocks(blocks, nBlocks)
+        raise ValueError("No blocks passed. Cannot determine links")
+
+    @staticmethod
+    def _getLinkedBlocks(
+        blocks: typing.Sequence[Block], nBlocks: int
+    ) -> dict[Block, AxialLink[Block]]:
+        # Use islice to avoid making intermediate lists of subsequences of blocks
+        lower = itertools.chain((None,), itertools.islice(blocks, 0, nBlocks - 1))
+        upper = itertools.chain(itertools.islice(blocks, 1, None), (None,))
+        links = {}
+        for low, mid, high in zip(lower, blocks, upper):
+            links[mid] = AxialLink(lower=low, upper=high)
```

Comment:
No problem! 

Disregarding the first and last block, we could write something like
```python
for i in range(1, len(blocks) - 1)
    below= blocks[i - 1]
    middle = blocks[i]
    above = blocks[i + 1]
    links[middle] = AxialLink(below, above)
```
And then the first and last block need special care for indicating nothing is below or above them. The `lower` and `upper` iterators take care of that. If they were lists, `lower[i] = links[blocks[i]].lower`, regardless of the location of `blocks[i]`.

So now we have three "iterables of blocks" between `lower`, `blocks`, and `upper`. `lower` and `upper` are not lists, but we probably could have made them with some list-slicing and the use of `chain`. I chose not to do list-slices because we just need to iterate over blocks, not all the functionality (and memory usage and construction time) of a list of blocks that slicing would produce